src/Common/memchrSmall.h

```diff
@@ -0,0 +1,46 @@
+#pragma once
+
+#include <bit>
+#include <Common/MemorySanitizer.h>
+#include <base/types.h>
+
+#if defined(__SSE2__)
+#    include <emmintrin.h>
+
+namespace detail
+{
+inline const char * memchrSmallAllowOverflow15Impl(const char * s, int c, ssize_t n)
+{
+    __msan_unpoison_overflow_15(s, n);
+
+    __m128i c16 = _mm_set1_epi8(c);
+    while (n > 0)
+    {
+        __m128i block = _mm_loadu_si128(reinterpret_cast<const __m128i *>(s));
+        UInt16 mask = _mm_movemask_epi8(_mm_cmpeq_epi8(block, c16));
+        if (mask)
+        {
+            auto offset = std::countl_zero(mask);
+            return offset < n ? s + offset : nullptr;
```

Comment:
Ah, I somehow missed this. This makes sure that `memchrSmallAllowOverflow15Impl` handles inputs with n < 16. The approach works but to be honest, we intentionally read memory beyond the end of the area and this requires unpoisoning (l. 14). Other memory safety tools (valgrind?) might rightfully still complain about invalid reads.

A conceptually nicer and safer approach would be to split the code into two phases:
- batch handling based on SSE2 (or wider SIMD instructions like AVX-512)
- tail handling.

Batch handling is called only if at least 16 bytes are left in the input.

EDIT: I just see that Common/memcpySmall.h performs similarly dirty tricks, and I assume the other code is also not Valgrind-safe. Okay, in that case, we can keep the logic here as is.